### DISCUSSION TITLE
Blog: Expose a Rust Library to Other Languages

### DISCUSSION THREAD

**Discussion by ogoffart** [slint-ui/slint#142] [Blog Comments]
Discussion thread for the blog post [Expose a Rust Library to Other Languages](https://sixtyfps.io/blog/expose-rust-library-to-other-languages.html)

https://sixtyfps.io/blog/expose-rust-library-to-other-languages.html

**Comment by matklad** [comment]
Great post, thanks! It reminded me of https://hsivonen.fi/modern-cpp-in-rust/ from a while ago.

I have a couple of questions: 

>The resulting library exports C mangled symbols.

Isn't this just unmangled? Or am missing some dark corner of C which requires mangling? 

>we wrap it an interface similar to std::span.

Can you use `std::span` directly? That is, does std::span have an ABI which can be relied upon by, eg, C?

**Reply by ogoffart** [reply]
Thanks @matklad 

> Isn't this just unmangled? Or am missing some dark corner of C which requires mangling?

I guess most people would say" unmangled" for C symbol indeed. In reality, the ABI is as usual a bit more complicated.
For example, on Windows, all C symbol gets preprended with a undescrore (`_`) https://docs.microsoft.com/en-us/cpp/error-messages/tool-errors/name-decoration?redirectedfrom=MSDN&view=msvc-160
And even on macOS there is something about having symbols starting with `\1`

> Can you use std::span directly?

As far as I know, `std::span`, like rust's slice does not have a specified ABI or data layout one can rely on (eg: the order of the pointers might depend on the standard library implementation)

**Comment by kassane** [comment]
Nice post! :smile: 

Recently, I heard from a friend about this project, which by the way seems quite promising.
Congratulations to all involved! :clap: 

Today I tried deliberately through the translator [C89 to Zig] embedded in the Zig compiler (still in development, generating a awful code) that ended up resulting in this:
- *C++ API (internals) to Zig:* https://gist.github.com/kassane/5b494d98061ce02ffa91da4110c56e4e -> **not tested yet!**

However, Zig does not follow the generic syntax of C++ and Rust or even support native [polymorphism and interfaces](https://github.com/ziglang/zig/issues/130), since the language focuses on C-like.
It would be interesting to have a generation of purely C headers for better interoperability integration facilitating access to any other language. An example is the [dotherside project](https://github.com/filcuc/dotherside) designed to use Qml/Qt5 binding in other languages.

**Reply by ogoffart** [reply]
If we were to generate bindings to Zig, we probably want to create pure C binding (as opposed to C++) from cbindgen. We would create another set of headers, where we give other flags to cbindgen, and do not add all the constructors/desturctors and such.
https://github.com/sixtyfpsui/sixtyfps/blob/488d71042e209e6f19f1d7b4f0df0c241ed1461f/tests/driver_lib/cbindgen.rs#L31

The problem with "lowering" to C is that C is rather primitive and we are loosing some information when doing the conversion.
Ideally there would be a "zigbindgen" which could generate zig header directly. But before this happens we would indeed have to generate some C that can go through the Zig translator, then manually doing the rest of the API.

**Reply by kassane** [reply]
I'm currently working with the possibility to generate zig code with cbindgen:
https://github.com/eqrion/cbindgen/pull/732

**Comment by Be-ing** [comment]
Interesting post! You may be interested to read how [libsignal-client does bindings from Rust to Node JS, Swift, and Java](https://community.signalusers.org/t/libsignal-client/28889/6).
